c10/cuda/CUDAStream.cpp

```diff
@@ -261,8 +261,10 @@ CUDAStream getStreamFromPool(
     const bool isHighPriority,
     DeviceIndex device_index) {
   initCUDAStreamsOnce();
-  if (device_index == -1)
+  if (device_index == -1) {
     device_index = current_device();
+    c10::cuda::SetTargetDevice();
```

Comment:
any directed test we could write here to catch this error directly, and not as "distributed tests are hanging"?

Reply:
Yes, I can write a C++ (or perhaps Python) test to catch something like this